src/promptflow-evals/promptflow/evals/evaluate/_evaluate.py

```diff
@@ -18,30 +25,40 @@ def _calculate_mean(df) -> Dict[str, float]:
     return mean_value.to_dict()
 
 
-def _validate_input_data_for_evaluator(evaluator, evaluator_name, data_df):
+def _get_missing_inputs(evaluator, columns):
     required_inputs = [
         param.name
         for param in inspect.signature(evaluator).parameters.values()
         if param.default == inspect.Parameter.empty and param.name not in ["kwargs", "args", "self"]
     ]
 
-    missing_inputs = [col for col in required_inputs if col not in data_df.columns]
-    if missing_inputs:
-        raise ValueError(f"Missing required inputs for evaluator {evaluator_name} : {missing_inputs}.")
+    return [col for col in required_inputs if col not in columns]
 
 
-def _validation(target, data, evaluators, output_path, tracking_uri, evaluation_name):
+def _validate_data(data):
     if data is None:
         raise ValueError("data must be provided for evaluation.")
 
-    if target is not None:
-        if not callable(target):
-            raise ValueError("target must be a callable function.")
-
     if data is not None:
         if not isinstance(data, str):
             raise ValueError("data must be a string.")
 
+
+def _validate_input_data_for_evaluator(evaluator, evaluator_name, columns, is_target_fn=False):
+    missing_inputs = _get_missing_inputs(evaluator, columns)
+    if missing_inputs:
+        if not is_target_fn:
+            raise ValueError(f"Missing required inputs for evaluator {evaluator_name} : {missing_inputs}.")
+        else:
+            raise ValueError(f"Missing required inputs for target : {missing_inputs}.")
+
+
+def _validation(target, columns, evaluators, output_path, tracking_uri, evaluation_name):
+
+    if target is not None:
+        if not callable(target):
+            raise ValueError("target must be a callable function.")
+

```

Comment:
In this code I wanted to avoid reading input data set multiple times. In the next lines of code we need to figure out, which columns were generated by target and which existed before.

Reply:
Can we load input dataset into dataframe in evaluate() function, then pass that dataframe to _validation() function for validation? In the later code of evaluate, you can also use that dataframe for checking which columns are generated by target.

Reply:
Yes, that was the main issue, I have tried to solve: First I was figuring out if the data is actually a path to the file in the `_validate_data` method, then I used to load it and read columns. The problem is that we still need to do the chaek before loading as data may be None or not a string.

Reply:
How about we repurpose `_validate_data` to `_validate_and_load_data`? In `_validate_and_load_data`, do the `data` None or empty check, loading the data into dataframe and return it. 

In this way,  dataframe is retrievable within evaluate() and we can pass it to _evaludation() same as before.

Reply:
Done